File: app/src/main/java/com/softwareoverflow/hiitultimate/database/dao/WorkoutDao.java
package com.softwareoverflow.hiitultimate.database.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.Transaction;

import com.softwareoverflow.hiitultimate.database.dto.Workout;
import com.softwareoverflow.hiitultimate.database.entity.WorkoutEntity;
import com.softwareoverflow.hiitultimate.database.entity.WorkoutSetEntity;

import java.util.List;

@Dao
public abstract class WorkoutDao implements BaseDao<WorkoutEntity> {

    // TODO - think about best methods for this given it won't have all it's data inside of it (necessarily)

    @Transaction
    @Query("Select * FROM workout")
    public abstract LiveData<List<Workout>> loadAllWorkouts();

    @Transaction
    @Query("SELECT * FROM workout WHERE workoutId = :workoutId")
    public abstract LiveData<Workout> loadWorkoutById(long workoutId);

    @Transaction
    public void createOrUpdate(WorkoutEntity workout, List<WorkoutSetEntity> sets,
                               WorkoutSetDao workoutSetDao) {
        long workoutId = createOrUpdate(workout);
        workout.setWorkoutId(workoutId);

        // New workout -> configure the WorkoutSetEntity foreign keys
        for (WorkoutSetEntity workoutSet : sets)
            workoutSet.setWorkoutId(workoutId);
        workoutSetDao.createOrUpdate(sets);
    }
}

File: app/src/main/java/com/softwareoverflow/hiit_trainer/data/dao/WorkoutDao.kt
package com.softwareoverflow.hiit_trainer.data.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.softwareoverflow.hiit_trainer.data.Workout
import com.softwareoverflow.hiit_trainer.data.entity.WorkoutEntity
import com.softwareoverflow.hiit_trainer.data.entity.WorkoutSetEntity

@Dao
interface WorkoutDao :
    BaseDao<WorkoutEntity> {

    @Query("SELECT * FROM Workout WHERE id = :workoutId")
    suspend fun getWorkoutById(workoutId: Long) : Workout

    @Query("SELECT * FROM Workout")
    fun getAllWorkouts(): LiveData<List<Workout>>

    @Query("SELECT COUNT(*) FROM workout")
    suspend fun getWorkoutCount(): Int

    @Transaction
    suspend fun createOrUpdate(workout: WorkoutEntity, workoutSets: List<WorkoutSetEntity>): Long {
        val id = createOrUpdate(workout)

        workoutSets.forEach {
            it.workoutId = id
        }

        createOrUpdateWorkoutSets(workoutSets)

        return id
    }

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun createOrUpdateWorkoutSets(workoutSets: List<WorkoutSetEntity>)

    @Query("SELECT workout.name FROM Workout workout JOIN WorkoutSet workout_set ON workout.id = workout_set.workoutId WHERE workout_set.exerciseTypeId = :exerciseTypeId")
    fun getExerciseTypeWorkoutNames(exerciseTypeId: Long) : List<String>
}